client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidationNearCacheConcurrencyTest.java

```diff
@@ -0,0 +1,101 @@
+   private static final String CACHE_NAME = InvalidationNearCacheConcurrencyTest.class.getName();
+
+   private boolean bloomFilter;
+
+   InvalidationNearCacheConcurrencyTest bloomFilter(boolean enabled) {
+      this.bloomFilter = enabled;
+      return this;
+   }
+
+   @Factory
+   public Object[] factory() {
+      return new Object[]{
+            new InvalidationNearCacheConcurrencyTest().bloomFilter(true),
+            new InvalidationNearCacheConcurrencyTest().bloomFilter(false),
+      };
+   }
+
+   @Override
+   protected String parameters() {
+      return "[bloomFilter = " + bloomFilter + "]";
+   }
+
+   @Override
+   protected EmbeddedCacheManager createCacheManager() throws Exception {
+      org.infinispan.configuration.cache.ConfigurationBuilder cb = hotRodCacheConfiguration();
+      EmbeddedCacheManager manager = TestCacheManagerFactory.createCacheManager(cb);
+      manager.createCache(CACHE_NAME, cb.build());
+      return manager;
+   }
+
+   @Override
+   protected ConfigurationBuilder createHotRodClientConfigurationBuilder(String host, int port) {
+      ConfigurationBuilder builder = super.createHotRodClientConfigurationBuilder(host, port);
+      builder.remoteCache(CACHE_NAME).nearCacheMode(NearCacheMode.INVALIDATED)
+            .nearCacheUseBloomFilter(bloomFilter)
+            .nearCacheMaxEntries(100);
+      return builder;
+   }
+
+   public void testConcurrentInvalidationWithRetrieval() throws InterruptedException, TimeoutException, ExecutionException {
+      RemoteCache<Integer, String> remoteCache = remoteCacheManager.getCache(CACHE_NAME);
+      remoteCache.put(1, "foo");
+      Cache<?, ?> cache = cacheManager.getCache(CACHE_NAME);
+      CheckPoint checkPoint = new CheckPoint();
+      checkPoint.triggerForever(Mocks.BEFORE_RELEASE);
+      // Block on the checkpoint when it is requesting segment 2 from node 2 (need both as different methods are invoked
+      // if the invocation is parallel)
+      InternalDataContainer<?, ?> realContainer = Mocks.blockingMock(checkPoint, InternalDataContainer.class, cache,
+            (stub, m) -> stub.when(m).peek(Mockito.anyInt(), Mockito.any()));
+
+      Future<Void> future = fork(() -> assertEquals(remoteCache.get(1), "foo"));
+
+      checkPoint.awaitStrict(Mocks.AFTER_INVOCATION, 10, TimeUnit.SECONDS);
+
+      // Put back the original so we don't block the remove below
+      TestingUtil.replaceComponent(cache, InternalDataContainer.class, realContainer, true);
+
+      // Remove the key
+      cache.keySet().forEach(cache::remove);
+
+      checkPoint.trigger(Mocks.AFTER_RELEASE);
+
+      // Need a sleep to guarantee the async remove event is sent to invalidate the near cache
+      // We can't use assertClient as other near cache tests do since we are blocking the thread above
+      Thread.sleep(500);
+
+      future.get(10, TimeUnit.SECONDS);
+
+      assertEquals(0, cache.size());
+
+      assertNull(remoteCache.get(1));
```

Comment:
as suggested by @jabolina , removing the sleep and replace this line with `eventuallyEquals(null, () -> remoteCache.get(1));` should do the trick. I don't understand why sleep is required.

Reply:
Changing it to `eventuallyEquals(null, () -> remoteCache.get(1));` should work, but cache.size does not as originally suggested.